### PR TITLE
feat(web): add ctrl+a / ctrl+d shortcuts to select / deselect all assets

### PR DESCRIPTION
**Pull request by Ethan13310**
This PR adds:

- Ctrl+A shortcut to select all assets.
- Ctrl+D shortcut to deselect all previously selected assets. It also cancels the ongoing "select all".
- "Select all" cancellation now works properly.
- "Select all" does not "freeze" the UI anymore if there is too many cached assets to select.

### CODE REVIEW

File: web/src/lib/components/photos-page/asset-grid.svelte [hunk 1]
```diff
@@ -202,12 +203,24 @@
...
 
-  const onKeyDown = (event: KeyboardEvent) => {
+  const onKeyDown = async (event: KeyboardEvent) => {
     if ($isSearchEnabled) {
       return;
     }
 
-    if (matchesShortcut(event, { key: 'Shift', shift: true })) {
+    // Select All
+    if (matchesShortcut(event, { key: 'A', ctrl: true })) {
```
Comment (michelheusschen) [inline web/src/lib/components/photos-page/asset-grid.svelte:212]: Why can't these be added to `shortcutList`?

File: web/src/lib/components/photos-page/asset-grid.svelte [hunk 2]
```diff
@@ -202,12 +203,24 @@
...
+      await selectAllAssets(assetStore, assetInteractionStore);
+    }
+    // Deselect All
+    else if (matchesShortcut(event, { key: 'D', ctrl: true })) {
+      event.preventDefault();
+      $isSelectingAllAssets = false;
+      assetInteractionStore.clearMultiselect();
+    }
+    // Start Multi-Selection
+    else if (event.key === 'Shift') {
```
Comment (michelheusschen) [inline web/src/lib/components/photos-page/asset-grid.svelte:223]: What's the reason behind this change?

File: web/src/lib/utils/asset-utils.ts [hunk 3]
```diff
@@ -224,6 +227,44 @@ export const getSelectedAssets = (assets: Set<AssetResponseDto>, user: UserRespo
   return ids;
 };
 
+export const selectAllAssets = async (assetStore: AssetStore, assetInteractionStore: AssetInteractionStore) => {
```
Comment (michelheusschen) [inline web/src/lib/utils/asset-utils.ts:230]: Maybe we can change `assetInteractionStore.selectAsset` to accept an array of assets and omit the second loop here altogether

File: web/src/lib/utils/asset-utils.ts [hunk 4]
```diff
@@ -224,6 +227,44 @@ export const getSelectedAssets = (assets: Set<AssetResponseDto>, user: UserRespo
...
+  if (get(isSelectingAllAssets)) {
+    // Selection is already ongoing
+    return;
+  }
+  isSelectingAllAssets.set(true);
+
+  try {
+    let selectedAssetCount = 1;
+    loadBuckets: for (const bucket of get(assetStore).buckets) {
```
Comment (michelheusschen) [inline web/src/lib/utils/asset-utils.ts:239]: Using `get` is unnecessary

File: web/src/lib/components/photos-page/asset-grid.svelte (hunk 2 above)
Comment (Ethan13310) [reply web/src/lib/components/photos-page/asset-grid.svelte]: Press Shift -> you enter in multi-selection mode
Then press Alt or Ctrl -> you are still in multi-selection mode (expected)
Then release Ctrl -> you are now stuck in multi-selection mode

Checking only the key code fixes that. I **know** there is a very low chance that someone does that, but that's still a (small) glitch waiting to happen

File: web/src/lib/utils/asset-utils.ts (hunk 3 above)
Comment (Ethan13310) [reply web/src/lib/utils/asset-utils.ts:230]: The second loop is useful to allow the UI to be refreshed during the "select all" process, with `delay(0)`

File: web/src/lib/components/photos-page/asset-grid.svelte (hunk 2 above)
Comment (michelheusschen) [reply web/src/lib/components/photos-page/asset-grid.svelte]: Ah makes sense, didn't think about that

File: web/src/lib/utils/asset-utils.ts (hunk 4 above)
Comment (jrasm91) [reply web/src/lib/utils/asset-utils.ts]: Isn't this a subscription? And this isn't a svelte component.

Comment (michelheusschen) [reply web/src/lib/utils/asset-utils.ts]: AssetStore is subscribable, but since we already have the value, using get is unnecessary

File: web/src/lib/utils/asset-utils.ts (hunk 3 above)
Comment (Ethan13310) [reply web/src/lib/utils/asset-utils.ts:230]: I just committed that change, but to be honest there is no perceptible performance improvement. Anyway that's still cleaner

File: web/src/lib/components/photos-page/asset-grid.svelte [hunk 5]
```diff
@@ -202,12 +205,17 @@
...
+    assetInteractionStore.clearMultiselect();
+  };
+
   const onKeyDown = (event: KeyboardEvent) => {
     if ($isSearchEnabled) {
       return;
     }
 
-    if (matchesShortcut(event, { key: 'Shift', shift: true })) {
+    if (event.key === 'Shift') {
```
Comment (alextran1502) [inline web/src/lib/components/photos-page/asset-grid.svelte:218]: Why do we reverting back to the previous implementation here?

Comment (michelheusschen) [reply web/src/lib/components/photos-page/asset-grid.svelte:218]: I asked that already and it handles edge cases better: https://github.com/immich-app/immich/pull/8105#discussion_r1532625879

**Review by michelheusschen — APPROVED** [review]
